Server side TypeScript support (#201)
graphile/graphile.github.io: src/pages/postgraphile/usage-schema.md

[graphql-js]: https://www.npmjs.com/package/graphql
[`pg-pool`]: https://www.npmjs.com/package/pg-pool

> # Server side TypeScript support

Reviewer: ```suggestion
# Server-side TypeScript support
```

> PostGraphile is taking care of the GraphQL API which different clients can use.

Reviewer: ```suggestion
PostGraphile takes care of building and serving a GraphQL API for various clients to use.
```

But it is not only possible to use the API from external clients. It is also

> possible to use the GraphQL API from within its own backend.

Reviewer: ```suggestion
But it is not only possible to use the API from external clients, it is also
possible to use the GraphQL API from within its own backend.
```

High-level overview:

> - use the graphql-code-generator to create TypeScript types for our GraphQL

Reviewer: ```suggestion
- use graphql-code-generator to create TypeScript types for our GraphQL
```

  schema
- use the generated types to query/mutate your data
- optional: use a Visual Studio Code extension to get IntelliSense

> ## TypeScript Code generation

Reviewer: ```suggestion
## TypeScript code generation
```

> We use the [{GraphQL} code generator](https://graphql-code-generator.com/) tools

Reviewer: ```suggestion
We use the [GraphQL code generator](https://graphql-code-generator.com/) tools
```

to create the TypeScript types in our backend code.

> The following `npm` packages will create typescript types:

Reviewer: ```suggestion
The following `npm` packages will create TypeScript types:
```

- `@graphql-codegen/cli` - the CLI tool to create the types

> - `@graphql-codegen/typescript` - create the typescript types. This is the main

Reviewer: ```suggestion
- `@graphql-codegen/typescript` - create the TypeScript types. This is the main
```

  package that we need

> - `@graphql-codegen/typescript-operations` - this input types for queries

Reviewer: ```suggestion
- `@graphql-codegen/typescript-operations` - generates types for queries/mutations/fragments
```

- optional: `@graphql-codegen/typescript-urql` - this produces typescript

>   queries from any custom `*.graphql` query/mutation files that we define.

Reviewer: ```suggestion
- optionally add a client specific generator: e.g. `@graphql-codegen/typescript-urql` 
```

Steps to get the backend typing support:

1. Start the development like described in the above section. Follow all the

>    steps to create your first GraphQL API with all needed schemas, tables,

Reviewer: ```suggestion
   steps to create your GraphQL API with all needed schemas, tables,
```

   roles, etc.
2. Configure in the PostGraphile setup to export the GraphQL schema:  

>    `exportGqlSchemaPath: './src/generated/schema.graphql'`

Reviewer: ```suggestion
2. Configure PostGraphile to export the GraphQL schema:  
   `exportGqlSchemaPath: './src/generated/schema.graphql'`
```

3. Start the project to let PostGraphile create the initial `schema.graphql`
   file. You can exclude this file from source control if you want to. But it

>    might be nice to see differences in the schema during check-ins.

Reviewer: ```suggestion
3. Start the project to let PostGraphile create the initial `schema.graphql`
   file. Since it's generated you can exclude this file from source control if
   you want to, but it is handy to see differences in the schema during check-ins
   and is useful for other tooling such as `eslint-plugin-graphql`.
```

4. Import the mentioned `npm` packages. You can find more plugins on their
   website in the [Plugins](https://graphql-code-generator.com/docs/plugins/)
   section.
5. Create a file `codegen.yml` in the root of your workspace.
   ```yaml
   overwrite: true
   schema: "./src/generated/schema.graphql"
   generates:
     # Creates the TypeScript types from the schema and any .graphql file
     src/generated/types.ts:
       documents: "src/**/*.graphql"
       plugins:
         - typescript
         - typescript-operations
         - typescript-urql
       config:
         withHOC: false
         withComponent: false
         withMutationFn: false
   config:
     scalars:
       DateTime: "string"
       JSON: "{ [key: string]: any }"
   ```
6. Run `graphql-codegen --config codegen.yml` to generate the types.
7. The generated types can now be used in your custom business logic code.

### Example

We have a movie table that we want to query from our backend system.

We can write a small GraphQL query file similar to this. It could be stored in

> `./src/graphql/getMovies.graphql`.

Reviewer: ```suggestion
`./src/graphql/getMovies.graphql`. (NOTE: this example uses the `@graphile-contrib/pg-simplify-inflector` plugin, your query might need to differ.)
```

```graphql
query GetMovies($top: Int!) {
  movies(first: $top) {
    nodes {
      id
      title
    }
  }
}
```

Save the file and run the code generation task.

Alternatively we can create some inline query directly in code like this:

```js
import { makeExtendSchemaPlugin, gql } from "graphile-utils";

// inside some function:

const GetMoviesDocument = gql`
  query Query {
    __typename
    movies(first: 3) {
      nodes {
        id
        title
      }
    }
  }
`;
```

The query can then be used in your code via the generated type or inline:

```js

  // This would be generated from the above .graphql file by the code generator
  // alternatively define it inline via gql` like shown above
  const doc = GetMoviesDocument;

>   const variables: GetMoviesQueryVariables = {

Reviewer: Please show this variable being imported

Reviewer 2: ok good point - will change it to a full makeExtendSchemaPlugin to see all the imports and definitions.

Author: ok good point - will change it to a full makeExtendSchemaPlugin to see all the imports and definitions.

    top: 3,
  };

  // Execute the query from your backend code
  // The "Query" type is your the root Query type. It contains all the query
  // endpoints from the GraphQL API with type support

>   const queryResult = await graphql.execute<Query>(

Reviewer: I don't think the type is correct here; I think this is used to describe the shape of the result which should be the TypeScript type that graphql-code-generator generated from the `GetMovies` operation via the `@graphql-codegen/typescript-operations` plugin:

```suggestion
  const queryResult = await graphql.execute<GetMovies>(
```

Author: Yes of course - thanks for pointing out. Should be `GetMoviesQuery`.

    resolveInfo.schema,
    doc,
    undefined,

>     context,

Reviewer: It's not clear where `context` has come from here - we should maybe wrap this code block in a function passing in the `context` and `resolveInfo`?

Reviewer 2: Same as above: will include this as a full example

Author: Same as above: will include this as a full exampl

    variables,
  );

  if (queryResult.errors) {

>     // do something in error case

Reviewer: ```suggestion
    // do something in error case
    throw queryResult.errors[0];
```

  }

  // the result can then be used to get the returned data
  const allTitles = queryResult.data?.movies?.nodes.map(
    (movie: any) => movie.title,

>   );

Reviewer: This whole code block is indented two spaces; if we wrap it in a function this is fine, but without that it should be outdented please :)

Author: Yes - include it in the full example.

```

### Usage in different resolvers

To use the GraphQL inside of our backend we need to get access to the "graphql"
variable. This variable is available for example in the following places:

```js
export const BusinessLogicPlugin = makeExtendSchemaPlugin((build: Build) => {
  const { graphql } = build;
```

```js
export const WrapResolverPlugin = makeWrapResolversPlugin({
  Mutation: {
    async updateMovieById(
      resolve: any,
      source: any,
      args: { [argName: string]: any },
      context: any,
      resolveInfo: GraphQLResolveInfo,
    ) {

>     const query = context.pgClient.query;

Reviewer: This doesn't show where to get `build.graphql` from... Also I'm not sure if it's possible to get `build` from Method 1 `makeWrapResolversPlugin` currently :flushed:

Author: Oh... I think I mixed that one with the PostgreSQL client. `build` is indeed not available. I think this would be a good addition to the context object to add the `graphql` object (in addition to the `pgClient`). For now, I will remove that section as the makeExtendSchemaPlugin is contained in the above example now.

```

### Build-time validation of inline documents

If you want to write inline
gql`queries it is possible to validate them during build-time. Adjust the created`codegen.yml`
file with the following section at the same level of the existing
"src/generated/types.ts:" part.

This will create types from any inline defined document. As we do not actually
need the files but only use them for build-time validation of our inline queries
we can have the types be put to some "trash" location.

```yaml
# Creates types for inline gql`...` defined queries.
# This is only used for build-time validation of those queries
trash/inlineGqlValidation.ts:
  documents: "src/**/*.ts"
  plugins:
    - typescript
    - typescript-operations
    - typescript-urql
  config:
    withHOC: false
    withComponent: false
    withMutationFn: false

> ```

Reviewer: I think we should delete this section. It's better to use something like `eslint-plugin-graphql` to do validation if you don't intend to use the types - it has better integration with editors/CI/etc.

```suggestion
```

Author: Thanks for the tip - will look into `eslint-plugin-graphql`.

## GraphQL IntelliSense

The above mentioned steps provide strong typing support. If you are developing
your code with VisualStudio Code you can get IntelliSense support both in

> .graphql files and in inline defined gql` template strings.

Reviewer: ```suggestion
.graphql files and in inline defined gql\` template strings.
```
